File: rust/vialen-core/src/raw_subscription/clash.rs
use super::{Node, Result};
use crate::config::{not_blank, protocols::int};
use serde_json::{Value, json};

// Java Map/List.toString is part of the accepted YAML scalar coercion contract.
pub(super) fn text(v: &Value) -> String {
    match v {
        Value::String(v) => v.clone(),
        Value::Array(a) => format!("[{}]", a.iter().map(text).collect::<Vec<_>>().join(", ")),
        Value::Object(o) => format!(
            "{{{}}}",
            o.iter()
                .map(|(k, v)| format!("{k}={}", text(v)))
                .collect::<Vec<_>>()
                .join(", ")
        ),
        _ => v.to_string(),
    }
}
fn scalar(v: &Value) -> Value {
    if v.is_null() {
        Value::Null
    } else {
        json!(text(v))
    }
}
fn join(v: &Value) -> Value {
    v.as_array()
        .map(|a| json!(a.iter().map(text).collect::<Vec<_>>().join("\n")))
        .unwrap_or(Value::Null)
}
fn number(v: &Value) -> Value {
    int(&text(v)).map_or(Value::Null, |n| json!(n))
}
fn required_number(v: &Value) -> Result<Value> {
    int(&text(v))
        .map(|n| json!(n))
        .ok_or("INVALID_CLASH_NUMBER")
}
fn is_true(v: &Value) -> bool {
    text(v) == "true"
}
fn map(v: &Value) -> impl Iterator<Item = (&String, &Value)> {
    v.as_object().into_iter().flat_map(|o| o.iter())
}
fn blank(v: &Value) -> bool {
    v.is_null() || !not_blank(&text(v))
}

pub(super) fn parse(root: Value) -> Result<Vec<Node>> {
    let proxies = root
        .get("proxies")
        .and_then(Value::as_array)
        .ok_or("INVALID_CLASH_PROXIES")?;
    let global = root
        .get("global-client-fingerprint")
        .map(text)
        .unwrap_or_default();
    let mut result = Vec::new();
    for proxy in proxies {
        let ty = proxy["type"].as_str().ok_or("INVALID_CLASH_TYPE")?;
        let kind = match ty {
            "socks5" => "SOCKS",
            "http" => "HTTP",
            "ss" => "Shadowsocks",
            "vmess" | "vless" => "VMess",
            "trojan" => "Trojan",
            "anytls" => "AnyTLS",
            "hysteria" | "hysteria2" => "Hysteria",
            "tuic" => "TUIC",
            _ => continue,
        };
        let mut b = json!({});
        match ty {
            "socks5" | "http" | "ss" => {
                b["serverAddress"] = json!(proxy["server"].as_str().ok_or("INVALID_CLASH_SERVER")?);
                b["serverPort"] = required_number(&proxy["port"])?;
                b["password"] = scalar(&proxy["password"]);
                b["name"] = scalar(&proxy["name"]);
                if ty == "ss" {
                    let cipher = proxy["cipher"].as_str().ok_or("INVALID_CLASH_CIPHER")?;
                    b["method"] = json!(if cipher == "dummy" { "none" } else { cipher });
                    let mut parts: Vec<String> = vec![];
                    if proxy.get("plugin").is_some() {
                        let opts = proxy["plugin-opts"]
                            .as_object()
                            .ok_or("INVALID_CLASH_PLUGIN")?;
                        let option = |key: &str| {
                            opts.get(key)
                                .filter(|v| !v.is_null())
                                .map(text)
                                .unwrap_or_default()
                        };
                        match proxy["plugin"].as_str() {
                            Some("obfs") => parts.extend([
                                "obfs-local".into(),
                                format!("obfs={}", option("mode")),
                                format!("obfs-host={}", option("host")),
                            ]),
                            Some("v2ray-plugin") => {
                                parts.extend([
                                    "v2ray-plugin".into(),
                                    format!("mode={}", option("mode")),
                                ]);
                                if option("mode") == "true" {
                                    parts.push("tls".into());
                                }
                                parts.extend([
                                    format!("host={}", option("host")),
                                    format!("path={}", option("path")),
                                ]);
                                if option("mux") == "true" {
                                    parts.push("mux=8".into());
                                }
                            }
                            _ => {}
                        }
                    }
                    b["plugin"] = json!(parts.join(";"));
                } else {
                    b["username"] = scalar(&proxy["username"]);
                    if ty == "http" {
                        b["security"] = json!(if is_true(&proxy["tls"]) { "tls" } else { "" });
                        b["sni"] = scalar(&proxy["sni"]);
                        b["allowInsecure"] = json!(is_true(&proxy["skip-cert-verify"]));
                    }
                }
            }
            "vmess" | "vless" | "trojan" => {
                if proxy["server"].is_null() || number(&proxy["port"]).is_null() {
                    continue;
                }
                b["serverAddress"] = scalar(&proxy["server"]);
                b["serverPort"] = number(&proxy["port"]);
                if ty == "vless" {
                    b["alterId"] = json!(-1);
                    b["packetEncoding"] = json!(2);
                }
                if ty == "trojan" {
                    b["security"] = json!("tls");
                }
                for (k, v) in map(proxy) {
                    match k.as_str() {
                        "name" => b["name"] = scalar(v),
                        "password" if ty == "trojan" => b["password"] = scalar(v),
                        "uuid" if ty != "trojan" => b["uuid"] = scalar(v),
                        "alterId" if ty != "trojan" && b["alterId"] != -1 => {
                            b["alterId"] = number(v)
                        }
                        "cipher" if ty != "trojan" && b["alterId"] != -1 => {
                            b["encryption"] = v.as_str().map_or(Value::Null, |s| json!(s))
                        }
                        "flow"
                            if ty != "trojan"
                                && b["alterId"] == -1
                                && v.as_str().is_some_and(|v| v.contains("xtls-rprx-vision")) =>
                        {
                            b["encryption"] = json!("xtls-rprx-vision")
                        }
                        "packet-encoding" if ty != "trojan" => {
                            b["packetEncoding"] = json!(match v.as_str() {
                                Some("packetaddr") => 1,
                                Some("xudp") => 2,
                                _ => 0,
                            })
                        }
                        "tls" if ty != "trojan" => {
                            b["security"] =
                                json!(if v.as_bool() == Some(true) { "tls" } else { "" })
                        }
                        "servername" | "sni" => b["sni"] = scalar(v),
                        "alpn" => b["alpn"] = join(v),
                        "skip-cert-verify" => b["allowInsecure"] = json!(v.as_bool() == Some(true)),
                        "client-fingerprint" => {
                            b["utlsFingerprint"] =
                                json!(v.as_str().ok_or("INVALID_CLASH_FINGERPRINT")?)
                        }
                        "reality-opts" => {
                            for (k, v) in map(v) {
                                b["security"] = json!("tls");
                                match k.as_str() {
                                    "public-key" => b["realityPubKey"] = scalar(v),
                                    "short-id" => b["realityShortId"] = scalar(v),
                                    _ => {}
                                }
                            }
                        }
                        "network" => match v.as_str() {
                            Some("h2" | "http") => b["type"] = json!("http"),
                            Some("ws" | "grpc") => b["type"] = v.clone(),
                            _ => {}
                        },
                        "ws-opts" => {
                            for (k, v) in map(v) {
                                match k.as_str() {
                                    "headers" => {
                                        for (k, v) in map(v) {
                                            if k.to_lowercase() == "host" {
                                                b["host"] = scalar(v);
                                            }
                                        }
                                    }
                                    "path" => b["path"] = scalar(v),
                                    "max-early-data" => b["wsMaxEarlyData"] = number(v),
                                    "early-data-header-name" => {
                                        b["earlyDataHeaderName"] = scalar(v)
                                    }
                                    "v2ray-http-upgrade" if v.as_bool() == Some(true) => {
                                        b["type"] = json!("httpupgrade")
                                    }
                                    _ => {}
                                }
                            }
                        }
                        "h2-opts" => {
                            for (k, v) in map(v) {
                                match k.as_str() {
                                    "host" => b["host"] = join(v),
                                    "path" => b["path"] = scalar(v),
                                    _ => {}
                                }
                            }
                        }
                        "http-opts" => {
                            for (k, v) in map(v) {
                                match k.as_str() {
                                    "path" => b["path"] = join(v),
                                    "headers" => {
                                        for (k, v) in map(v) {
                                            if k.to_lowercase() == "host" {
                                                if !v.is_array() {
                                                    return Err("INVALID_CLASH_HEADERS");
                                                }
                                                b["host"] = join(v);
                                            }
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }
                        "grpc-opts" => {
                            for (k, v) in map(v) {
                                if k == "grpc-service-name" {
                                    b["path"] = scalar(v);
                                }
                            }
                        }
                        "smux" => {
                            for (k, v) in map(v) {
                                match k.as_str() {
                                    "enabled" => b["enableMux"] = json!(is_true(v)),
                                    "max-streams" => b["muxConcurrency"] = required_number(v)?,
                                    "padding" => b["muxPadding"] = json!(is_true(v)),
                                    _ => {}
                                }
                            }
                        }
                        "ech-opts" => {
                            for (k, v) in map(v) {
                                if k == "enable" {
                                    b["enableECH"] = json!(is_true(v));
                                }
                            }
                        }
                        _ => {}
                    }
                }
            }
            "anytls" | "hysteria" | "hysteria2" | "tuic" => {
                let hy = ty.starts_with("hysteria");
                let mut hop = String::new();
                let mut ip = String::new();
                if hy {
                    b["protocolVersion"] = json!(if ty == "hysteria" { 1 } else { 2 });
                }
                for (k, v) in map(proxy) {
                    if v.is_null() {
                        continue;
                    }
                    let k = k.replace('_', "-");
                    match k.as_str() {
                        "name" => b["name"] = scalar(v),
                        "server" => {
                            if ty != "tuic" && !v.is_string() {
                                return Err("INVALID_CLASH_SERVER");
                            }
                            b["serverAddress"] = scalar(v);
                        }
                        "port" => {
                            if hy {
                                b["serverPorts"] = scalar(v);
                            } else {
                                b["serverPort"] = required_number(v)?;
                            }
                        }
                        "ports" if hy => hop = text(v),
                        "password" => {
                            b[if ty == "tuic" {
                                "token"
                            } else if ty == "hysteria2" {
                                "authPayload"
                            } else if ty == "anytls" {
                                "password"
                            } else {
                                continue;
                            }] = scalar(v)
                        }
                        "sni" => b["sni"] = scalar(v),
                        "skip-cert-verify" => b["allowInsecure"] = json!(is_true(v)),
                        "client-fingerprint" if ty == "anytls" => {
                            b["utlsFingerprint"] =
                                json!(v.as_str().ok_or("INVALID_CLASH_FINGERPRINT")?)
                        }
                        "alpn" if ty != "hysteria2" => {
                            b["alpn"] = join(v);
                            if ty == "hysteria" && b["alpn"].is_null() {
                                b["alpn"] = json!("h3");
                            }
                        }
                        "obfs" if ty == "hysteria" => b["obfuscation"] = scalar(v),
                        "obfs-password" if ty == "hysteria2" => b["obfuscation"] = scalar(v),
                        "auth-str" if ty == "hysteria" => {
                            b["authPayloadType"] = json!(1);
                            b["authPayload"] = scalar(v);
                        }
                        "up" | "down" if hy => {
                            b[if k == "up" {
                                "uploadMbps"
                            } else {
                                "downloadMbps"
                            }] = json!(
                                int(text(v).split(' ').next().unwrap_or(""))
                                    .unwrap_or(if ty == "hysteria" { 100 } else { 0 })
                            )
                        }
                        "recv-window-conn" | "recv-window" if ty == "hysteria" => {
                            b[if k == "recv-window-conn" {
                                "connectionReceiveWindow"
                            } else {
                                "streamReceiveWindow"
                            }] = json!(int(&text(v)).unwrap_or(0))
                        }
                        "disable-mtu-discovery" if ty == "hysteria" => {
                            b["disableMtuDiscovery"] = json!(is_true(v) || text(v) == "1")
                        }
                        "ip" if ty == "tuic" => ip = text(v),
                        "token" if ty == "tuic" => {
                            b["protocolVersion"] = json!(4);
                            b["token"] = scalar(v);
                        }
                        "uuid" if ty == "tuic" => b["uuid"] = scalar(v),
                        "disable-sni" if ty == "tuic" => b["disableSNI"] = json!(is_true(v)),
                        "reduce-rtt" if ty == "tuic" => b["reduceRTT"] = json!(is_true(v)),
                        "congestion-controller" if ty == "tuic" => {
                            b["congestionController"] = scalar(v)
                        }
                        "udp-relay-mode" if ty == "tuic" => b["udpRelayMode"] = scalar(v),
                        _ => {}
                    }
                }
                if not_blank(&hop) {
                    b["serverPorts"] = json!(hop);
                }
                if not_blank(&ip) {
                    b["serverAddress"] = json!(ip);
                    if blank(&b["sni"]) && !super::is_ip(&ip) {
                        b["sni"] = json!(ip);
                    }
                }
            }
            _ => unreachable!(),
        }
        if matches!(ty, "http" | "vmess" | "vless" | "trojan") {
            if b["security"] == "tls"
                && blank(&b["sni"])
                && !blank(&b["host"])
                && !super::is_ip(&text(&b["host"]))
            {
                b["sni"] = b["host"].clone();
            }
            if !blank(&b["realityPubKey"]) && blank(&b["utlsFingerprint"]) {
                b["utlsFingerprint"] = json!(if not_blank(&global) {
                    &global
                } else {
                    "chrome"
                });
            }
        }
        result.push(Node::new(kind, b, true));
    }
    Ok(result)
}

File: core/importer/clash.go
package importer

import (
	"encoding/json"
	"github.com/killertop/Vialen/core/profile"
	"math"
	"strconv"
	"strings"
)

func parseClash(text string) (Result, error) {
	out := result("clash")
	v, e := parseYAML(text)
	if e != nil {
		return out, e
	}
	root, e := object(v)
	if e != nil {
		return out, bad("INVALID_CLASH", "Expected a Clash mapping")
	}
	list, ok := root["proxies"].([]any)
	if !ok {
		return out, bad("INVALID_CLASH", "Expected a proxies array")
	}
	if e = sourceLimit(len(list)); e != nil {
		return out, e
	}
	rootFields := fields{m: root}
	global := rootFields.str("global-client-fingerprint")
	if rootFields.err != nil {
		return out, rootFields.err
	}
	for i, v := range list {
		p, e := clashProfile(v, global)
		out.add(i, p, e)
	}
	return out, nil
}
func clashProfile(v any, global string) (profile.Profile, error) {
	m, e := object(v)
	if e != nil {
		return profile.Profile{}, e
	}
	f := fields{m: m}
	if f.has("dialer-proxy") {
		dependency := f.str("dialer-proxy")
		if f.err != nil {
			return profile.Profile{}, f.err
		}
		if dependency != "" {
			if strings.TrimSpace(dependency) == "" {
				return profile.Profile{}, fieldError("dialer-proxy")
			}
			return profile.Profile{}, bad("DEPENDENT_NODE", "Proxy dependencies require an explicit chain")
		}
	}
	// These constraints cannot be represented by a standalone profile. Do not
	// silently turn a constrained connection into an unconstrained one.
	for _, key := range []string{"interface-name", "fingerprint", "certificate", "private-key", "certificate-path", "private-key-path"} {
		if key == "private-key" && strings.EqualFold(f.str("type"), "wireguard") {
			continue
		}
		if f.has(key) {
			value := f.str(key)
			if f.err != nil {
				return profile.Profile{}, f.err
			}
			if value != "" {
				return profile.Profile{}, bad("UNSUPPORTED_SECURITY_FIELD", "Unsupported connection constraint: "+key)
			}
		}
	}
	if f.has("routing-mark") {
		mark := f.uint("routing-mark", math.MaxUint32)
		if f.err != nil {
			return profile.Profile{}, f.err
		}
		if mark != 0 {
			return profile.Profile{}, bad("UNSUPPORTED_SECURITY_FIELD", "Unsupported connection constraint: routing-mark")
		}
	}
	ty := strings.ToLower(f.str("type"))
	p := profile.Profile{Name: f.str("name"), Server: f.str("server"), Port: uint16(f.uint("port", 65535))}
	switch ty {
	case "socks5", "socks4", "socks4a":
		p.Type = "socks"
		p.Socks = &profile.Socks{Version: strings.TrimPrefix(ty, "socks"), Username: f.str("username"), Password: f.str("password")}
	case "http":
		p.Type = "http"
		p.HTTP = &profile.HTTP{Username: f.str("username"), Password: f.str("password")}
	case "ss":
		p.Type = "shadowsocks"
		p.Shadowsocks = &profile.Shadowsocks{Method: f.str("cipher"), Password: f.str("password"), Plugin: f.str("plugin")}
		if f.has("plugin-opts") {
			opts := f.child("plugin-opts")
			plugin := p.Shadowsocks.Plugin
			switch plugin {
			case "obfs":
				p.Shadowsocks.Plugin = "obfs-local"
				p.Shadowsocks.PluginOptions = "obfs=" + opts.str("mode") + ";obfs-host=" + opts.str("host")
			case "v2ray-plugin":
				parts := []string{}
				if s := opts.str("mode"); s != "" {
					parts = append(parts, "mode="+s)
				}
				if opts.boolean("tls") {
					parts = append(parts, "tls")
				}
				if s := opts.str("host"); s != "" {
					parts = append(parts, "host="+s)
				}
				if s := opts.str("path"); s != "" {
					parts = append(parts, "path="+s)
				}
				if opts.boolean("mux") {
					parts = append(parts, "mux=8")
				}
				p.Shadowsocks.PluginOptions = strings.Join(parts, ";")
			default:
				opts.err = bad("UNSUPPORTED_PLUGIN", "Unsupported Shadowsocks plugin options")
			}
			f.accept(opts)
		}
	case "vmess":
		p.Type = "vmess"
		p.VMess = &profile.VMess{UUID: f.str("uuid"), Security: f.str("cipher"), AlterID: uint32(f.uint("alterId", math.MaxUint32)), PacketEncoding: f.str("packet-encoding")}
	case "vless":
		p.Type = "vless"
		p.VLESS = &profile.VLESS{UUID: f.str("uuid"), Flow: f.str("flow"), PacketEncoding: f.str("packet-encoding")}
	case "trojan":
		p.Type = "trojan"
		p.Trojan = &profile.Trojan{Password: f.str("password")}
	case "hysteria2", "hy2":
		p.Type = "hysteria2"
		hy := &profile.Hysteria2{Password: f.str("password"), UpMbps: clashMbps(&f, "up"), DownMbps: clashMbps(&f, "down")}
		if f.has("ports") {
			hy.ServerPorts = parsePortRanges(f.str("ports"))
		}
		if s := f.str("obfs"); s != "" {
			hy.Obfs = &profile.Obfs{Type: s, Password: f.str("obfs-password")}
		}
		p.Hysteria2 = hy
	case "tuic":
		p.Type = "tuic"
		p.TUIC = &profile.TUIC{UUID: f.str("uuid"), Password: f.str("password"), CongestionControl: f.str("congestion-controller"), UDPRelayMode: f.str("udp-relay-mode"), ZeroRTTHandshake: f.boolean("reduce-rtt")}
	case "shadowtls":
		p.Type = "shadowtls"
		p.ShadowTLS = &profile.ShadowTLS{Version: uint32(f.uint("version", 3)), Password: f.str("password")}
	case "anytls":
		p.Type = "anytls"
		p.AnyTLS = &profile.AnyTLS{Password: f.str("password")}
	case "wireguard":
		p.Type = "wireguard"
		wg := &profile.WireGuard{PrivateKey: f.str("private-key"), PublicKey: f.str("public-key"), PreSharedKey: f.str("pre-shared-key"), MTU: uint32(f.uint("mtu", math.MaxUint32)), PersistentKeepalive: uint32(f.uint("persistent-keepalive", 65535)), AllowedIPs: f.strings("allowed-ips")}
		for _, k := range []string{"ip", "ipv6"} {
			if s := f.str(k); s != "" {
				if !strings.Contains(s, "/") {
					if k == "ip" {
						s += "/32"
					} else {
						s += "/128"
					}
				}
				wg.Address = append(wg.Address, s)
			}
		}
		wg.Reserved = reserved(&f, "reserved")
		p.WireGuard = wg
	default:
		return p, bad("UNSUPPORTED_PROTOCOL", "Unsupported Clash proxy type")
	}
	advancedFields(&f, &p, true)
	if p.Type != "wireguard" {
		tls := &profile.TLS{DisableSNI: f.boolean("disable-sni"), Enabled: f.boolean("tls"), ServerName: f.str("sni"), Insecure: f.boolean("skip-cert-verify"), ALPN: f.strings("alpn"), Fingerprint: f.str("client-fingerprint")}
		if tls.ServerName == "" {
			tls.ServerName = f.str("servername")
		}
		if tls.Fingerprint == "" {
			tls.Fingerprint = global
		}
		if p.Type == "shadowtls" || p.Type == "trojan" || p.Type == "hysteria2" || p.Type == "tuic" || p.Type == "anytls" {
			tls.Enabled = true
		}
		if f.has("reality-opts") {
			r := f.child("reality-opts")
			tls.Reality = &profile.Reality{PublicKey: r.str("public-key"), ShortID: r.str("short-id")}
			tls.Enabled = true
			f.accept(r)
		}
		if f.has("ech-opts") {
			e := f.child("ech-opts")
			tls.ECH = &profile.ECH{Enabled: e.boolean("enable"), Config: listable(e, "config")}
			f.accept(e)
		}
		if tls.Enabled {
			p.TLS = tls
		}
		transport, e := clashTransport(&f)
		if e != nil {
			return p, e
		}
		p.Transport = transport
	}
	if f.err != nil {
		return p, f.err
	}
	server, e := profile.NormalizeServer(p.Server)
	if e != nil {
		return p, bad("INVALID_SERVER", "Invalid proxy server")
	}
	p.Server = server
	return p, nil
}
func clashMbps(f *fields, k string) uint32 {
	if !f.has(k) {
		return 0
	}
	if _, ok := f.m[k].(json.Number); ok {
		return uint32(f.uint(k, math.MaxUint32))
	}
	s, ok := f.m[k].(string)
	if !ok {
		f.err = fieldError(k)
		return 0
	}
	s = strings.TrimSpace(s)
	if strings.HasSuffix(strings.ToLower(s), "mbps") {
		s = strings.TrimSpace(s[:len(s)-4])
	}
	n, e := strconv.ParseUint(s, 10, 32)
	if e != nil {
		f.err = fieldError(k)
		return 0
	}
	return uint32(n)
}
func parsePortRanges(s string) []string {
	if strings.TrimSpace(s) == "" {
		return nil
	}
	items := strings.Split(s, ",")
	for i := range items {
		items[i] = strings.ReplaceAll(strings.TrimSpace(items[i]), "-", ":")
	}
	return items
}
func reserved(f *fields, k string) []uint32 {
	if !f.has(k) {
		return nil
	}
	a, ok := f.m[k].([]any)
	if !ok {
		f.err = fieldError(k)
		return nil
	}
	out := []uint32{}
	for _, v := range a {
		entry := fields{m: map[string]any{"value": v}}
		out = append(out, uint32(entry.uint("value", 255)))
		if entry.err != nil {
			f.err = fieldError(k)
		}
	}
	return out
}
func clashTransport(f *fields) (*profile.Transport, error) {
	network := f.str("network")
	if network == "" || network == "tcp" {
		return nil, f.err
	}
	t := &profile.Transport{Type: network}
	switch network {
	case "ws":
		opts := f.child("ws-opts")
		t.Path = opts.str("path")
		t.MaxEarlyData = uint32(opts.uint("max-early-data", math.MaxUint32))
		t.EarlyDataHeaderName = opts.str("early-data-header-name")
		if opts.boolean("v2ray-http-upgrade") {
			t.Type = "httpupgrade"
		}
		if opts.has("headers") {
			h := opts.child("headers")
			for k := range h.m {
				if strings.EqualFold(k, "host") {
					t.Host = []string{h.str(k)}
				} else {
					if t.Headers == nil {
						t.Headers = map[string][]string{}
					}
					t.Headers[k] = []string{h.str(k)}
				}
			}
			opts.accept(h)
		}
		f.accept(opts)
	case "h2":
		t.Type = "http"
		opts := f.child("h2-opts")
		t.Host = opts.strings("host")
		t.Path = opts.str("path")
		f.accept(opts)
	case "http":
		opts := f.child("http-opts")
		paths := opts.strings("path")
		if len(paths) > 1 {
			return nil, bad("UNSUPPORTED_FIELD", "Multiple HTTP transport paths are not supported")
		}
		if len(paths) == 1 {
			t.Path = paths[0]
		}
		if opts.has("headers") {
			h := opts.child("headers")
			for k := range h.m {
				if strings.EqualFold(k, "host") {
					t.Host = h.strings(k)
				} else {
					if t.Headers == nil {
						t.Headers = map[string][]string{}
					}
					t.Headers[k] = h.strings(k)
				}
			}
			opts.accept(h)
		}
		f.accept(opts)
	case "grpc":
		opts := f.child("grpc-opts")
		t.ServiceName = opts.str("grpc-service-name")
		f.accept(opts)
	case "httpupgrade":
		opts := f.child("http-upgrade-opts")
		t.Path = opts.str("path")
		if host := opts.str("host"); host != "" {
			t.Host = []string{host}
		}
		f.accept(opts)
	default:
		return nil, bad("UNSUPPORTED_TRANSPORT", "Unsupported transport type")
	}
	return t, f.err
}
